### DISCUSSION TITLE
Feature request: HPA on controllers

### DISCUSSION THREAD

**Discussion by ericjohansson89** [fluxcd/flux2#2312] [Proposals]
Hello!

First I would like that we are using Flux more and more and think it is an excellent product.

We have started to come to a part where we are running Flux in multiple clusters with varying load in flux through the amount of Flux Custom Resources in the cluster. It feels like setting individual overrides per cluster for resources is an unwanted pattern. Rather I would like the controllers to scale horizontally based on the load.

The work could be delegated between the workers from the leader to split the load. I understand that the source controller might be hard to scale horizontally but all other controllers should be able to split the workload?

**Comment by kingdonb** [comment]
This is not at all straightforward to implement from today, as Flux controllers are not at present designed to be horizontally scaled.

They all implement the same leader election protocol as far as I understand, which means that no matter how you scale the deployments with an HPA, only one instance of a controller will be managing its associated CRD resources at any given time.

You can however set `--concurrency` to increase the number of resources that a Flux controller will process in parallel (and this comes with a performance penalty in the form of increased memory usage, mostly, for the controllers that handle objects which have a large memory footprint like Helm Controller and Source Controller, and probably Kustomize Controller as well.)

You will also need to increase the memory limit if you increased `--concurrency`. This is vertical scaling which is not what you asked for, but it is what Flux supports for now at least. Hope this helps!

**Reply by stefanprodan** [reply]
If the Kubernetes controller-runtime will support sharding we could reconsider this, but until then, scaling using Go routines (`--concurrency`) is the only way.

**Reply by ericjohansson89** [reply]
I am afraid I do not have enough knowledge about kubernetes controller-runtime. But why would it not be possible for the elected leader to share workload to the other controller pods? I just feels like a big anti-pattern to set individual resource quotas for each cluster based on the amount of CRDs it has in the cluster. I would be much more convenient to set to run between 1-6 controller pods that share the workload which then means we could use same setting across our clusters. And about sharding, how are other tools (for example prometheus-operator) handling sharding?

Really glad for the feedback and discussion! 👍

**Reply by ericjohansson89** [reply]
@kingdonb

**Reply by kingdonb** [reply]
> why would it not be possible for the elected leader to share workload to the other controller pods

I think we're saying that it makes sense for implementation of a feature like sharding to happen at the controller-runtime level.

That's a feature that should be supported in the runtime, and according to this search nobody else has asked for it yet: https://github.com/kubernetes-sigs/controller-runtime/search?q=sharding

Similar results for kubebuilder project. I think the general sense I am getting is that if controller-runtime has not implemented it yet and if controller-runtime doesn't support sharding, we do not want to undertake to implement it ourselves outside of the runtime.

I might be overstepping because it's really not my decision, it's up to the consensus of the maintainers, but if you feel strongly about this maybe it will help to make a case with benchmarks?

One of the issues the Flux team struggles with is, we do not have large enough clusters of our own that we can effectively prove performance points and make the case that SSA provided amazing performance boosts, or whatever. We really don't have the numbers to support with data about whether Flux performs well or not at the boundaries of scale or have any idea of real limits of what Flux can do. If you have clusters that are hitting performance bottlenecks, we should get the details into an issue first, and support with data as much as possible how the current approach fails or what makes it inadequate.

I always thought the control plane would fall over first, and that when scaling Flux to immense proportions I would find the clusters actually fall over before Flux does when it comes to performance, but like I say I don't really have a way to spin up a 1000 node cluster and test at the boundary of what a single Flux pod can do when scaled up with excessive memory limits and CPU unthrottled, and I am just learning to use Prometheus myself. I don't know if I am quite ready where I could even approach making such a benchmark study, (but I've spent a fair amount of time learning the basics now.)

I appreciate very much you taking the time to report this and I wonder if you've followed through to the Flux monitoring guide, can provide some screenshots or other empirical data to show what limitations you're experiencing and where the limits are at right now? There is a `Flux Cluster Stats` and `Flux Control Plane` grafana dashboard that has information about how many Flux reconcilers and manifest sources you have, how many are failing, and how long they all spend on average. If you have access to this resource, it will help to provide that type of data to understand better. ❓

**Reply by ericjohansson89** [reply]
Thank for your answer. We are probably nowhere near those size yet but we are moving more and more things over to our clusters to be delivered by flux. As things progress we could probably give you more data. However that is not the thing I am requesting it is the fact that we have many clusters which vary in size. And I do not want to configure each cluster individually on resources, I would like that to be handled by horizontal scaling.

**Reply by kingdonb** [reply]
Got it. My suggestion for now is to monitor flux, follow the [monitoring guide](https://fluxcd.io/docs/guides/monitoring/), and also follow the [webhook receiver guide](https://fluxcd.io/docs/guides/webhook-receivers/) to maximize performance of Flux.

Flux's performance is a funny thing because even when it's configured properly, chances are good that most of Flux's time is spent idle cycling, while changes are not being pushed, so the best way to measure it effectively must take into account first and foremost any reflection of performance that affects the user's experience.

We discussed in (https://github.com/fluxcd/flux2/discussions/2069) a freshness SLO, and there's a link to a deeper discussion I had about how Flux's reliability can be measured. The central conclusion though is that: spending performance budget on Flux idle cycling at a high rate is generally unnecessary if you can use webhooks to notify the Flux `GitRepository` resources when they should sync. You should use webhooks, so that you can dial back the `spec.interval` of Flux resources, so that idle cycling does not substantially negatively impact the performance as Flux scales to thousands of resources managed.

Flux can keep up with lots of changes from many different sources, and we provide detailed monitoring dashboards so you can verify the performance meets expectations of your developer/ops teams.

The concern is noted, and your intention is understood. Flux (in the future) could adapt its own configuration to reflect any necessary performance adjustments, whether that comes in the form of an HPA, or dynamic concurrency setting and a VPA, or something else, is an implementation detail. Thank you for using Flux and for engaging here!

**Reply by matheuscscp** [reply]
@kingdonb is this still the current status of sharding/horizontal scaling, i.e. not supported?

**Reply by matheuscscp** [reply]
we are currently deciding whether we deploy Flux in-cluster for each cluster, or deploy a single Flux instance inside an infra cluster managing everything else in the non-infra clusters. if horizontal scaling is not supported yet then I think a single Flux will lead to problems in a couple of years according to our growth expectations

**Reply by kingdonb** [reply]
The `--watch-all-namespaces` flag is there, so you can use it to isolate multiple instances of Flux from one another, but I would need to see performance of what you're describing and check if the guidance we provide has any effect before I recommend that anti-pattern.

If your growth expectations are scaling to the point where a single Flux pod with high concurrency as the document about scaling describes, and you've tried this advice already:

https://fluxcd.io/docs/cheatsheets/bootstrap/#increase-the-number-of-workers
... and it isn't helping, then we can take a look at it.

But if your problem is purely hypothetical, I doubt that you will find a single Flux pod can't be scaled up to handle whatever your Kubernetes control plane can also handle. It's a shared bottleneck according to my understanding, so it seems more likely to me in your hypothetical scaling situation that you overloaded the Kubernetes control plane itself, rather than up past what Flux can handle.

Can you say more about the specific expectations, and whether you have any experimental data that suggests Flux is the bottleneck that can't be scaled past by simply adding more memory/CPU and increasing the concurrency as described in the doc? It's easy to make educated guesses, but hard data are a lot more actionable.

**Reply by matheuscscp** [reply]
actually we are at a very early stage with Flux and dont have any data yet, but we do have an important security decision for our current CD tool which is having it only in a separate infra cluster, and deploy all our apps across all our clusters from there. this cluster has more isolation, the other clusters cant initiate connections to it (only the other way around is allowed in the firewall), and this cluster is the only one with network connectivity to our git service at the moment.

we want to replace our current CD tool with Flux, so the first question is: can we keep this architecture/security decision?

I know that Flux supports this architecture with the obvious additional cost in operational complexity (configuring service accounts for external use in each cluster, etc.), which is probably a cost we are willing to pay. but how does it scale in terms of performance? if it turns out that sharding/horizontal scaling is not supported, I'm not sure I want to even try and see the scale limits for N apps x M clusters, and would rather recommend a different approach to my team from the beginning

**Reply by kingdonb** [reply]
I'm afraid we are tilting at windmills without data to support ourselves, and I've been somewhat confused about which thread this question came from and which problem it was aimed at solving. Sorry for any confusion. There are two different questions, and they don't necessarily have the same answer. Managing many instances of Flux on the same cluster is not generally supported, and if you're looking for guidance about how to build your Flux installation according to best practices, the advice is to scale with the concurrency model of Flux, which is not horizontal pod autoscalers, but adding more goroutines and enough memory/CPU to support them, for the service of additional concurrency.

You should not add more pods to Flux controllers because Flux will do leader election, and then only one controller pod will be effective while the others wait in "hot-standby" mode in case they are the next elected leader. If you want to run multiple instances of Flux on the same cluster disabling this leader election, that is not supported by Controller Runtime upstream and it seems unlikely to be supported soon. Flux would certainly not implement this support for sharding controllers ahead of Controller Runtime, which is a standard for Kubernetes controllers.

1) Can you run Flux on one cluster, multiple times, to shard and support different workloads independently on the same cluster. – Yes, you can, but it's not recommended to disable the leader election, and I wouldn't design my large installation of Flux this way. I do not have large swarths of thousand clusters with 10 Flux Kustomizations each to test with so I can't give specific performance numbers. I'm not sure at what point the scale would break down. The pull model is generally recommended instead, for another reason – so there is no single point of failure or "main cluster with access to every other cluster" that could be compromised.

This SPOF seems like a much riskier thing to have than allowing clusters to access the Git repo without going through a sort of "home office cluster" with a trusted/privileged position first. Installing Flux onto each cluster and letting the definition for what workloads to install on the cluster should live on that cluster is a generally better approach, and I think it resolves the scaling question that you had.

2) Can you manage many clusters from one central instance of Flux – yes, you can. 

Using the Flux `Kustomization.spec.kubeconfig`, you can manage many Flux clusters from one central cluster, even prior to installing Flux on those target clusters. This is the pattern that we use in our enterprise product too: Cluster-API provides a cluster-admin secret that our (enterprise Weave GitOps) Cluster Bootstrap Controller uses to bootstrap another instance of Flux on the target cluster, along with some base configuration, usually to bootstrap the support for your cluster's RBAC at a minimum, (and whatever else).

Then from that point on there is no SPOF or "regular updates" that flow from any central cluster. The central cluster can go down, and updates continue to flow to the clusters without any centralized point of failure. This is neither in conflict with the idea of keeping central control nor the different idea of decentralized operation, both goals are met. An important factor for scaling is that now, heavy lifting can be done on the target clusters and each cluster serves as a natural boundary for the "sharding" of Flux. The main Flux cluster is only responsible for sending instructions to every target cluster, probably only `flux-system` and some Flux Kustomization resources there; and the remainder of the scale problem is solved naturally because each cluster has only its own local workloads to manage, and not some tree of many clusters.

Besides bootstrapping from the management cluster, each Flux handles the workload on its own cluster, and no Flux instance needs to wait very long to update itself, or to scale to outrageous proportion to manage 10k clusters. Once for a brief while (at the time they are first being bootstrapped) is likely the only time that our management cluster needs to operate on all 10k clusters at once. Even during upgrades, you can avoid the central bottleneck and push out updates to all clusters simultaneously, or in batches, whatever makes the most sense for you. The source of truth all comes from Git.

Since target clusters are not required to have any write access to Git, and at no point after bootstrap are the target clusters required to receive instructions from a central cluster; we think this is more secure as a pattern and more "GitOps-y". You can make it even more secure through Cosign verification and OCI artifacts; that way your releases are verified with some policy before they get sent out to the clusters, and the verification can be performed again in-situ on the clusters that are deploying the verified workloads.

**Reply by ghost** [reply]
> Can you run Flux on one cluster, multiple times, to shard and support different workloads independently on the same cluster. – Yes, you can

@kingdonb how exactly is this possible at the moment? what "multiple times" means exactly?

**Comment by stefanprodan** [comment]
Support for horizontal scaling of all Flux controllers with sharding based on label selectors is coming, see https://github.com/fluxcd/pkg/pull/525
